Add pre-defined CASMO energy group structures to openmx.mgxs module

wbinventor opened the pull request:
This PR takes a first stab at addressing #1104 by introducing CASMO's pre-defined energy group structures to `openmc.mgxs`. The group structures are the same as those I used for my thesis and corresponding publications. This feature is implemented as a static `EnergyGroups.get_energy_groups(name)` class method where `name` takes on the name of the desired group structure. As an example, this now allows for the following:

```python
casmo_2group = openmc.mgxs.EnergyGroups.get_energy_groups("CASMO-2")
casmo_8group = openmc.mgxs.EnergyGroups.get_energy_groups("CASMO-8")
casmo_70group = openmc.mgxs.EnergyGroups.get_energy_groups("CASMO-70")
```

nelsonag commented on openmc/mgxs/groups.py:
```diff
@@ -77,6 +77,69 @@ def group_edges(self, edges):
         self._group_edges = np.array(edges)
 
+    @staticmethod
+    def get_energy_groups(name):
```
Is there a more instructive name for this? Perhaps `get_common_energy_groups`?

wbinventor replied:
Yes, I like that name better as well.

nelsonag commented on openmc/mgxs/groups.py:
```diff
@@ -77,6 +77,69 @@ def group_edges(self, edges):
         cv.check_greater_than('number of group edges', len(edges), 1)
         self._group_edges = np.array(edges)
 
+    @staticmethod
```
I am curious about your thoughts as to why you went with this factory approach instead of adding these to the within the model subpackage.

wbinventor replied:
I don't have a strong reason -- I felt it was a natural pattern to use here, and it mirrors the factory pattern used for the `MGXS.get_mgxs(...)` static class method used by the `openmc.mgxs.Library` class to build `MGXS` instances for different reaction types and domains.

nelsonag commented on openmc/mgxs/groups.py:
```diff
@@ -77,6 +77,69 @@ def group_edges(self, edges):
+        ----------
+        name : str
+            The name of the energy group structure. Pre-defined structures
+            include "CASMO-X" (where X is 2, 4, 8, 16, 25, 40 or 70).
```
Is there a reference you can cite for the CASMO structures?

wbinventor replied:
Unfortunately, there is not a reference :-( I had a CASMO user plug in a special card into his CASMO input file in order to output the group structures. They are all derived from WIMS-69. I've added in a reference to the CASMO code itself; let me know if you like that, want it removed, or have another idea altogether.

smharper commented:
I have no comments on this code, but I just want to say welcome back, @wbinventor :)  It's good to have you back in the PR tab!

wbinventor commented:
@paulromano I'm also not a "huge fan" of using a staticmethod per se, but also don't have a strong opinion either way and agree that it this feature is quite hidden as is. How would you suggest that these not be hidden away - would this be a dictionary attached to the `openmc` or `openmc.mgxs` modules? One thing to consider is that this dictionary would increase (albeit only slightly) the time necessary to `import openmc.mgxs` since it will need to initialize each of the common `EnergyGroups` objects. Also, these objects would (albeit slightly) bloat the `openmc.mgxs` namespace. The only benefit I see to using a dictionary is that it would require fewer characters to type; but perhaps I am missing something? As for what data type this should feature should return to the user, it's worth noting that the numpy array is easily accessible via `EnergyGroups.group_edges`. 

For what it's worth, I see this feature as very similar to the [`sklearn.datasets`](http://scikit-learn.org/stable/modules/classes.html#module-sklearn.datasets) and [`torchvision.models`](https://pytorch.org/docs/stable/torchvision/models.html#torchvision-models) which are commonly used dataset and model loading APIs for ML/DL. Neither uses a staticmethod nor a dictionary, but rather specially named functions. This makes sense for their respective applications since they don't return values of a single data type as will be the case for us, but it also makes this feature of each package much more visible than a factory method would, while not bloating the namespace. Anyway, I just wonder if there is anything we might learn from how those teams have implemented this.

Notwithstanding all of what I just said, I don't have a strong opinion on this, so of course I'll revise to whatever you and @nelsonag and the users settle on!

paulromano commented:
@wbinventor I'm thinking maybe a dictionary in the `openmc.mgxs` namespace. Perhaps:
```Python
energies = openmc.mgxs.GROUP_STRUCTURE['CASMO-8']

# Using the energies
efilter = openmc.EnergyFilter(energies)
groups = openmc.mgxs.EnergyGroups(energies)
```
Anyone else have thoughts on this? @openmc-dev/committers

paulromano commented:
Also, for what it's worth, import time really shouldn't be an issue. Python still has to instantiate all the floating point constants either way, and in the case of the staticmethod, they are permanently stored on the code object for the function itself:
```python
>>> import openmc.mgxs
>>> openmc.mgxs.EnergyGroups.get_common_energy_groups.__code__.co_consts
('Returns one of many commonly used energy group structures.\n\n        Pre-defined structures include "CASMO-X" (where X is 2, 4, 8, 16, 25,\n        40 or 70) from the [CASMO]_ lattice physics code.\n\n        References\n        ----------\n        .. [CASMO] J. Rhodes, K. Smith, and D. Lee, "CASMO-5 Development and\n           Applications", *Proc. ANS Topical Meeting on Reactor Physics* (2006).\n\n        Parameters\n        ----------\n        name : str\n            The name of the energy group structure.\n\n        Returns\n        -------\n        EnergyGroups\n            The energy group structure \n\n        ', 'CASMO-2', 0.0, 0.625, 20000000.0, 'CASMO-4', 5530.0, 821000.0, 'CASMO-8', 0.058, 0.14, 0.28, 4.0, 'CASMO-16', 0.03, 0.35, 0.85, 0.972, 1.02, 1.097, 1.15, 1.3, 'CASMO-25', 1.855, 9.877, 15.968, 148.73, 9118.0, 111000.0, 500000.0, 1353000.0, 2231000.0, 3679000.0, 6065500.0, 'CASMO-40', 0.015, 0.042, 0.08, 0.1, 0.18, 0.22, 0.95, 1.5, 2.1, 2.6, 3.3, 27.7, 48.052, 'CASMO-70', 0.005, 0.01, 0.02, 0.025, 0.035, 0.05, 0.067, 0.25, 0.3, 0.32, 0.4, 0.5, 0.78, 0.91, 9.96e-06, 1.045, 1.071, 1.123, 75.501, 367.26, 0.09069, 1425.1, 2239.5, 3519.1, 9118.3, 15030.0, 24780.0, 40850.0, 67340.0, 183000.0, 302500.0, 'Energy group structure "{}" is not recognized')
```

GiudGiud commented:
I think it's a great idea to have all the common group structures ready for users, and I'll be using it. 

With regards to implementation, both ways work just as well in my opinion.
However, no user will know about this unless this is used in an example and added to the documentation

johnnyliu27 commented on openmc/mgxs/groups.py:
```diff
@@ -77,6 +77,76 @@ def group_edges(self, edges):
+                1.11e5, 5.e5, 8.21e5, 1.353e6, 2.231e6, 3.679e6, 6.0655e6,
+                2.e7])
+        elif name == "CASMO-70":
+            group_edges = np.array([
```
Just found three values may be typos when comparing the 70-group structure with the one I often use. I think the correct values should be:
```
group_edges_70[30] = 9.96e-1
group_edges_70[51] = 9.069e2
group_edges_70[56] = 9.118e3
```
All the other values are the same as I checked.
Again, welcome back! @wbinventor

wbinventor replied:
Thanks for double checking the values @johnnyliu27! I'll update the values this evening. Oddly enough, I do think that you and I are either using different structures, or the typo originated in the structure I used for my thesis. Which version of CASMO did your structures come from? I know that the structures changed slightly between v4 and v5; mine (supposedly) came from v5.

wbinventor commented:
@paulromano -- I'm there with you now on making this a dictionary on the `openmc.mgxs` namespace and will update this tonight. Per @GiudGiud's comment, do you have any advice on how to comment/expose the dictionary to make it visible in the Sphinx documentation?

wbinventor commented:
I updated this PR to put the CASMO group structures as NumPy arrays in a `dict` in the `openmc.mgxs` namespace. I also fixed three typos per @johnnyliu27's comment. 

In addition, I made an effort to raise the profile of this feature in the `openmc.mgxs` docs per @GiudGiud's comment by introducing the new `docs/_templates/mymodulevar.rst` Sphinx template file. This file is now used by `docs/pythonapi/mgxs.rst` file to document the `GROUP_STRUCTURES` alongside the `EnergyGroup` class in the Python API docs. Unfortunately, I didn't figure out how to completely expose my docstring for `GROUP_STRUCTURES` to Sphinx. The following is a snippet of the output from `make html` which shows that the reference for CASMO is not found:

```shell
...
.../openmc/docs/source/pythonapi/mgxs.rst:15:<autosummary>:1: WARNING: citation not found: CASMO
...
```

Any ideas on how to properly comment the `GROUP_STRUCTURES` module variable?

paulromano commented on openmc/mgxs/__init__.py:
```diff
@@ -1,4 +1,43 @@
+
+"""
+
+GROUP_STRUCTURES['CASMO-2'] = np.array([
```
One more thing before we merge: add `import numpy as np` at the top of this module to not rely on transitive imports.

wbinventor replied:
Done

paulromano commented:
Thanks again for the contribution; should be useful to many people!
